xen/arch/arm/cpufreq/scpi_cpufreq.c

```diff
@@ -0,0 +1,353 @@
+#include <xen/err.h>
+#include <xen/cpufreq.h>
+#include <asm/bug.h>
+#include <asm/percpu.h>
+
+#include "scpi_protocol.h"
+
+extern struct device *get_cpu_device(unsigned int cpu);
+
+struct scpi_cpufreq_data
+{
+    struct processor_performance *perf;
+    struct cpufreq_frequency_table *freq_table;
+    struct scpi_dvfs_info *info;
+    int domain_id;
+};
+
+static struct scpi_cpufreq_data *cpufreq_driver_data[NR_CPUS];
+
+static struct cpufreq_driver scpi_cpufreq_driver;
+
+static struct scpi_ops *scpi_ops;
+
+static unsigned int scpi_cpufreq_get(unsigned int cpu)
+{
+    struct scpi_cpufreq_data *data;
+    struct cpufreq_policy *policy;
+    const struct scpi_opp *opp;
+    int idx;
+
+    if ( cpu >= nr_cpu_ids || !cpu_online(cpu) )
+        return 0;
+
+    policy = per_cpu(cpufreq_cpu_policy, cpu);
+    if ( !policy || !( data = cpufreq_driver_data[policy->cpu] ) )
+        return 0;
+
+    /* Sanity check */
+    if ( data->domain_id < 0 || !data->info )
+        return 0;
+
+    idx = scpi_ops->dvfs_get_idx(data->domain_id);
+    if ( idx < 0 )
+        return 0;
+
+    opp = data->info->opps + idx;
+
+    return opp->freq / 1000; /* kHz */
+}
+
+static unsigned int check_freq(unsigned int cpu, unsigned int freq)
+{
+    unsigned int curr_freq;
+    int i;
+
+    for ( i = 0; i < 100; i++ )
+    {
+        curr_freq = scpi_cpufreq_get(cpu);
+        if ( curr_freq == freq )
+            return 1;
+        udelay(10);
+    }
+    return 0;
+}
+
+static int scpi_cpufreq_target(struct cpufreq_policy *policy,
+                               unsigned int target_freq, unsigned int relation)
+{
+    struct scpi_cpufreq_data *data = cpufreq_driver_data[policy->cpu];
+    struct processor_performance *perf;
+    struct cpufreq_freqs freqs;
+    cpumask_t online_policy_cpus;
+    unsigned int next_state = 0; /* Index into freq_table */
+    unsigned int next_perf_state = 0; /* Index into perf table */
+    unsigned int j;
+    int result = 0;
+    const struct scpi_opp *opp;
+    int idx, max_opp;
+
+    if ( unlikely(!data) || !data->perf || !data->freq_table )
+        return -ENODEV;
+
+    perf = data->perf;
+    result = cpufreq_frequency_table_target(policy,
+                                            data->freq_table,
+                                            target_freq,
+                                            relation, &next_state);
+    if ( unlikely(result) )
+        return -ENODEV;
+
+    cpumask_and(&online_policy_cpus, &cpu_online_map, policy->cpus);
+
+    next_perf_state = data->freq_table[next_state].index;
+    if ( perf->state == next_perf_state )
+    {
+        if ( unlikely(policy->resume) )
+            policy->resume = 0;
+        else
+            return 0;
+    }
+
+    freqs.old = perf->states[perf->state].core_frequency * 1000;
+    freqs.new = data->freq_table[next_state].frequency;
+
+    /* Sanity check. TODO Put this at the beginning. */
+    if ( data->domain_id < 0 || !data->info )
```

Comment:
This is like `assert()`? Or this is like "invalid configuration"?

Reply:
Like assert(). I think I can even drop "data->domain_id < 0" check.

Reply:
I have spotted such checks in different places. Looks like they can be converted to `assert()`. Or removed at all.